### DISCUSSION TITLE
Retry arbitrary blocks with single context manager using AST

### DISCUSSION THREAD

**Discussion by CNSeniorious000** [hynek/stamina#86] [Ideas]
In the [docs](https://stamina.hynek.me/en/latest/tutorial.html#arbitrary-code-blocks):

> Since iterators can’t catch exceptions and context managers can’t execute the same block multiple times, we need to combine them to achieve that.

Maybe it is possible to get the ast of the caller's frame? So maybe this feature can br achieved more elegantly?

**Comment by hynek** [comment]
While I would never put such an abomination into production, I would be really curious to look at that. 😂

I find the two indentations annoying too, but start to rewrite ASTs would break the mental model people have of the behavior of a Python application and make it a net negative. It would also confuse all kinds of tooling like type checkers and linters, etc.

But still, I would love to see a PoC. :D
